keps/1136-provisioning-request-support/README.md

```diff
@@ -0,0 +1,185 @@
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+Introduce an (AdmissionCheck)[https://github.com/kubernetes-sigs/kueue/tree/main/keps/993-two-phase-admission]
+that will use (`ProvisioningRequest`)[https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/proposals/provisioning-request.md]
+to ensure that there is enough capacity in the cluster before 
+admitting a workload.
+
+## Motivation
+
+Currently Kueue admits workloads based on the quota check alone. 
+This works reasonably well in most cases, but doesn't provide 
+guarantee that an admitted workload will actually schedule 
+in full in the cluster. With `ProvisioningRequest`, SIG-Autoscaling owned
+(ClusterAutoscaler)[https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler]
+opens a way for stronger (but still not hard-guaranteed) all-or-nothing
+scheduling in an autoscaled cloud environment. 
+
+Before admission, CA will check whether there is enough resources and
+provide them if their number is not sufficient (details
+depend on the exact engine used with `ProvisioningRequest)`.
+
+### Goals
+
+* Provide Kueue integration with `ProvisioningReqest`.
+* Define how users can configure what Kueue puts into `ProvisioningRequest`.
+* Provide a list of test cases to ensure that all corner cases are covered.
+
+### Non-Goals
+
+* Define how Cluster Autoscaler handles ProvisioningRequest.
+* Define underling cloud-specific behavior.
+
+## Proposal
+
+* Introduce a new controller in Kueue that will act as AdmissionCheck based on
+  the status of created `ProvisioningRequest`.
+
+* Introduce a new cluster-scoped CRD to configure how `ProvisioningRequest` should be used.
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+I want to admit workloads only after ClusterAutoscaler running on my cloud provider
+expands a dedicated node group on which the workload will be run.
+
+#### Story 2
+
+I want to admit workloads only after a CheckCapacity request to ClusterAutoscaler
+succeeds.
+
+### Risks and Mitigations
+
+There doesn't seem to be much risks or mitigations. (Two phase admission process)[
+was added specifically for use cases like this.
+
+## Design Details
+
+The new ProvisioningRequest controller will:
+
+* Watch for all workloads that require an `AdmissionCheck` with controller
+name set to `"ProvisioningRequestController"`. For that it will also need to
+to watch all `AdmissionCheck` definitions to understand whether the particular
+check is in fact `ProvisioningRequest` or not.
+
+* For each of such workloads create a `ProvisioningRequest` (and accompanying
+PodTemplates) requesting capacity for all podsets from the workload. Annotate
+the workload with the ProvisioningRequest name under:
+`"cluster-autoscaler.kubernetes.io/consume-provisioning-request"`
+To understand what details should it put into `ProvisioningRequest` the controller
+will also need to watch `ProvisioningRequestConfigs`.
```

Comment:
The relevant fields (ProvisioningClassName and Parameters) are immutable in the `ProvisioningRequest`  definition, should the `ProvisioningRequest`  be recreated or let to "run" with the old parameters?

If the `ProvisioningRequestConfig` is deleted should we  `ProvisioningRequest`  or let it run until the CA is taking it into a  final state.

If we choose "run" for both of them , we don't need to watch for `ProvisioningRequestConfig` changes , we only need to be able to read then upon `ProvisioningRequest` creation.

Reply:
The circumstances change (for whatever reason) and the ProvisioningRequest is not valid anymore, it should be immediately deleted.

Reply:
let's add it to the KEP